### ISSUE TITLE
Not detecting newly added files

### ISSUE THREAD

**Issue by pensierinmusica** [gruntjs/grunt-contrib-watch#282]
If I understand correctly, `grunt-contrib-watch` should detect any added, modified, or deleted file that matches the patters specified in the Grunt configuration file.

On the other hand, I noticed that in my case it doesn't detect newly added files at all.
Here's an example of my Grunt configuration:

```
  grunt.initConfig({
    watch: {
      options: {
        livereload: true
      },
      styles: {
        files: ['<%= yeoman.app %>/styles/sections/*.styl'],
        tasks: ['stylus', 'autoprefixer'],
        options: {livereload: false}
      },
      iconfont: {
        files: ['<%= yeoman.app %>/images/icons/*.svg'],
        tasks: ['webfont', 'stylus'],
      }
```

`grunt-contrib-watch` perfectly detects modifications to existing files, but for example if I add a new `.styl` file in the specified folder nothing happens. I'm using Node 0.10.24 on OSX Mavericks.

Any advice on what might be causing the problem? 

Thanks  :)

**Comment by rotemdan** [comment]
I experienced the same problem when I had "./" at the beginning of the search path. I removed it and it now it does detect newly added files in non-empty directories. I can confirm that it still doesn't detect new files in an empty directory. Applying @Denney's suggestion didn't seem to help. 

(node ```0.12```, grunt: ```0.4.5```, grunt-contrib-watch: ```0.6.1``` on Windows 7 64bit)

As this issue has been unresolved for more than a year now (and more than year and a half for similar ones such as #166). Being the most popular (and one of the most useful) grunt plugins I find it weird no solution has been found yet and considering looking for an alternative.

**Comment by dominikwilkowski** [comment]
:+1:

**Comment by UnknownHero** [comment]
https://github.com/unbalanced/grunt-simple-watch 
Fix problem for me

**Comment by foxx** [comment]
I think this may be related to Vagrant shared folders (due to lack of inotify?)

**Comment by tandrewnichols** [comment]
FWIW, this task also does not work for added (or deleted, I think) files if the file paths are absolute.

**Comment by roNn23** [comment]
https://github.com/unbalanced/grunt-simple-watch 
fixed the problem for me also!

**Comment by believer-ufa** [comment]
It's a really important issue for comfortable development, and yes, **grunt-simple-watch** fixed this problem for me too.

**Comment by leedorian** [comment]
@UnknownHero It's works for me, this issue waste me 2 days, OMG.

**Comment by ezk84** [comment]
Having the same issue where newly created less files aren't watched.

I've managed to reproduce this with a minimal example `Gruntfile.js`:

```javascript
module.exports = function (grunt) {
    grunt.loadNpmTasks('grunt-contrib-watch');

    var config = {
        watch: {
            less: {
                files: ['src/**/*.less'],
                tasks: ['listfiles:less']
            },
            js: {
                files: ['src/**/*.js'],
                tasks: ['listfiles:js']
            }
        },
        listfiles: {
            js: {
                src: ['src/**/*.js']
            },
            less: {
                src: ['src/**/*.less']
            }
        }
    };

    grunt.initConfig(config);

    grunt.registerMultiTask('listfiles', 'List files matched', function () {
        console.log(this.filesSrc);
    });
};
```

Now if we create a `src/test.js` and then launch `grunt watch`:
 - Creating a `src/test2.js` triggers the watch
 - Creating a `src/test.less` does not trigger the watch.

What seems to be happening is that the absence of at least one matching file in the directory prevents `watch` from triggering on subsequent changes in that directory (for a given pattern). This is why empty directories are completely ignored, but also non-empty directories that don't have any matching files.

Hope this can help bring about a solution.

**Comment by joskuijpers** [comment]
+1

**Comment by benjaminfunk** [comment]
+1

**Comment by rocket13531** [comment]
+1

**Comment by mcela** [comment]
+1

**Comment by toptalo** [comment]
:+1:

**Comment by gfpacheco** [comment]
Still no solution or explanation for this issue?

**Comment by joskuijpers** [comment]
Nope, still having this problem. I will try to debug myself tomorrow. It is a really annoying bug.

**Comment by seomantix** [comment]
It seems that there is only a problem with new files, if the watcher is configured to look at specific file types. I now use 'src/js/**' instead of 'src/js/**/*.js' and the watcher works as expected. The watcher is triggered after adding new files and after changing them without to restart the watch task. Only a little workaround with some restrictions but maybe helpful for debugging.

**Comment by dr0bz** [comment]
> I think this may be related to Vagrant shared folders (due to lack of inotify?)

@foxx in my case with vagrant and nfs it is really the problem with lack of inotify. https://github.com/unbalanced/grunt-simple-watch solves it through polling but it uses 20%-25% cpu of the guest machine instead of 5% with normal watch.

**Comment by dr0bz** [comment]
grunt-simple-watch lacks livereload support :(

**Comment by Rybadour** [comment]
@seomantix I tried to remove the file extension and the "*" but it still doesn't work. Perhaps when you tested it you had an old file where you didn't expect?

**Comment by zachrybaker** [comment]
Still happening.  OSX

**Comment by taburetkin** [comment]
still have this issue, win10.
and i have found that
if i have file pattern `files:['tmpl/**/*.html']`
and i have `tmpl/folderA` when watch started
now i want to create `tmpl/folderB` and `tmpl/folderB/new.html`
1. creating folder `tmpl/folderB`
2. creating new file in folder `tmpl/folderB` , just to be sure that watcher fails to rise event
3. then creating  any new file in `tmpl/folderA` (because it existed on watcher start and watcher will rise event which will fix something)
4. and after that creating the `tmpl/folderB/new.html`. done. watcher will rise event.

i think something bad happens after new folder is created. )

**Comment by wrager** [comment]
The issue also exists in this case:

```
styles: {
    // Adding files is detected
    files: [['lib/**/*.less']],
    tasks: ['...']
},
devInfo: {
    // Adding files is not detected
    files: [['lib/**/index.ts']],
    tasks: '...'
},
options: {
    spawn: false,
    event: ['changed', 'added', 'deleted']
}
```

**Comment by axten** [comment]
I can confirm that the bug still exists.
It only occurs if this file type doesn't exist in this folder before adding.
When adding a second file watch works as expected.

**Comment by ve3** [comment]
grunt-cli v1.3.2
grunt v1.0.4
2019-06-12
problem still exists.
